### PR TITLE
Switch to bb8

### LINKED ISSUE #669: [spanner] Make DbPool fully async
Spanner's pool `get` still goes through a threadpool. It could potentially be a source of "clogged" connections when the server's overloaded. We saw similar "clogging" under overloaded conditions when all Db calls went through a thread pool.

Take a scenario where there's slowness/hiccups between syncstorage and Spanner:

- actix-threadpool defaults to a size of num_cpus * 5 (e.g. 40)
- we have a larger DB_POOL_MAX_SIZE (e.g. a value of 300)
- syncstorage-rs is overloaded with a large number of pending pool `get` requests

The threadpool could potentially "backup" with pending requests, impart because we're overloaded w/ requests (> 300 at a time) and also impart due to Spanner taking a long time to respond to the Pool's `is_valid` calls.

The tasks "backed up" in the threadpool may become stale (that is their corresponding db connection may become stale) by the time they are actually gotten to and ran by the threadpool.

### PR DESCRIPTION
**Pull request by fzzzy**
## Description

r2d2 is a blocking database pool, but the rest of our code is now fully nonblocking. Switch to bb8, which has a very similar interface. The fact that bb8 needs to know the lifetime necessitates a large number of changes, but they are fairly mechanical.

Also, @Mcat12 switched some handlers to async/await, which helps give bb8 the lifetime it needs; database connections now are acquired in the handler instead of in an extractor, which requires the 'static lifetime.

## Testing

This should be a transparent change; all tests should pass, things should continue working as expected.

## Issue(s)

Closes #669.

### CODE REVIEW

File: src/db/spanner/models.rs [hunk 1]
```diff
@@ -1609,38 +1609,38 @@ impl SpannerDb {
     }
 }
 
-unsafe impl Send for SpannerDb {}
+unsafe impl Send for SpannerDb<'_> {}
```
Comment (Mcat12) [inline src/db/spanner/models.rs:1612]: We can remove this unsafe Send impl and the Send requirement from the Db trait. Tested and it compiles/runs.

File: src/db/spanner/manager.rs [hunk 2]
```diff
@@ -70,24 +88,17 @@ impl ManageConnection for SpannerConnectionManager {
...
-            use_test_transactions: false,
+            use_test_transactions: self.test_transactions,
         })
     }
 
-    fn is_valid(&self, conn: &mut Self::Connection) -> Result<(), Self::Error> {
+    async fn is_valid(&self, conn: Self::Connection) -> Result<Self::Connection, Self::Error> {
         let mut req = GetSessionRequest::new();
         req.set_name(conn.session.get_name().to_owned());
         if let Err(e) = conn.client.get_session(&req) {
```
Comment (Mcat12) [inline src/db/spanner/manager.rs:98]: Replace `get_session` with `get_session_async`.

File: src/server/mod.rs [hunk 3]
```diff
@@ -63,7 +63,7 @@ macro_rules! build_app {
             .wrap(ErrorHandlers::new().handler(StatusCode::NOT_FOUND, ApiError::render_404))
             // These are our wrappers
             .wrap(middleware::precondition::PreConditionCheck::new())
-            .wrap(middleware::db::DbTransaction::new())
+            // .wrap(middleware::db::DbTransaction::new())
```
Comment (jrconlin) [inline src/server/mod.rs:66]: Should we kill this or is my overwhelming paranoia starting to rub off on y'all?

File: src/web/extractors.rs [hunk 4]
```diff
@@ -758,15 +756,12 @@ impl FromRequest for CollectionPostRequest {
...
-                    CollectionParam,
-                    BsoQueryParams,
-                    BsoBodies,
-                )>::from_request(&req, &mut payload)
-                .await?;
+
+            let user_id = HawkIdentifier::from_request(&req, &mut payload).await?;
+            let collection = CollectionParam::from_request(&req, &mut payload).await?;
+            let query = BsoQueryParams::from_request(&req, &mut payload).await?;
+            let mut bsos = BsoBodies::from_request(&req, &mut payload).await?;
```
Comment (jrconlin) [inline src/web/extractors.rs:763]: +100
The other is probably more efficient around stack use, but this is far more clear.

File: src/web/middleware/mod.rs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-pub mod db;
+// pub mod db;
```
Comment (jrconlin) [inline src/web/middleware/mod.rs:1]: kill ?

File: src/web/middleware/precondition.rs [hunk 6]
```diff
@@ -3,20 +3,19 @@ use std::{cell::RefCell, rc::Rc};
...
 
+use crate::db::transaction::DbTransactionPool;
 use actix_web::{
     dev::{Service, ServiceRequest, ServiceResponse, Transform},
     http::{header, StatusCode},
-    Error, HttpMessage, HttpResponse,
+    Error, FromRequest, HttpMessage, HttpResponse,
 };
-use futures::future::{self, Either, FutureExt, LocalBoxFuture, TryFutureExt};
+use futures::future::{self, FutureExt, LocalBoxFuture};
```
Comment (jrconlin) [inline src/web/middleware/precondition.rs:18]: Love the fact that we don't have to juggle the return values anymore and can kill `Either`

File: src/web/middleware/precondition.rs [hunk 7]
```diff
@@ -76,140 +75,116 @@ where
...
+            let transaction_pool = DbTransactionPool::extract(&req).await?;
+            let resource_ts = transaction_pool
+                .transaction(|db| async move {
+                    db.extract_resource(user_id, collection, bso_opt)
+                        .await
+                        .map_err(Into::into)
+                })
+                .await?;
+            let sreq = ServiceRequest::from_parts(req, payload)
+                .unwrap_or_else(|_| panic!("TODO: this should never happen"));
```
Comment (jrconlin) [inline src/web/middleware/precondition.rs:152]: Isn't there an `unreachable!()` macro for this?

File: src/server/mod.rs (hunk 3 above)
Comment (Mcat12) [reply src/server/mod.rs:66]: Oh ya, I meant to delete that after commenting it out. There's also #693

File: src/web/extractors.rs (hunk 4 above)
Comment (Mcat12) [reply src/web/extractors.rs:763]: Related: #698

File: src/web/middleware/precondition.rs (hunk 7 above)
Comment (Mcat12) [reply src/web/middleware/precondition.rs:152]: Ya... I'm pretty sure this should never happen, but I also wanted to leave it as a TODO: #695

File: src/server/mod.rs (hunk 3 above)
Comment (fzzzy) [reply src/server/mod.rs:66]: Leave it for the followup bug. This pr is big enough as is.

File: src/web/middleware/mod.rs (hunk 5 above)
Comment (fzzzy) [reply src/web/middleware/mod.rs:1]: Do this as part of #693

File: src/db/transaction.rs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+        async move {
+            let no_agent = HeaderValue::from_str("NONE")
+                .expect("Could not get no_agent in DbTransactionPool::from_request");
+            let useragent = req
+                .headers()
+                .get("user-agent")
+                .unwrap_or(&no_agent)
+                .to_str()
+                .unwrap_or("NONE");
+            info!(">>> testing db middleware"; "user_agent" => useragent);
```
Comment (pjenvey) [inline src/db/transaction.rs:160]: should really kill this while we're here (but it can wait for a separate PR)

**Comment by fzzzy** [comment]
Great, thanks Phil. We opened a few followup bugs, but thinking about this I think it's pretty important for every request to only do one transaction. So I think we should refactor the middlewares that use their own transaction into functions before this lands.

https://github.com/mozilla-services/syncstorage-rs/issues/714

https://github.com/mozilla-services/syncstorage-rs/issues/714

I guess I'm going to have to look at the code in a little more detail to see exactly what should be done.

**Comment by fzzzy** [comment]
It looks like I pasted the link to 714 twice in my previous comment. I meant to paste this:

https://github.com/mozilla-services/syncstorage-rs/issues/710

Although this issue won't be fixed using the method suggested in the issue, Phil's suggested way of avoiding the double-transaction accomplishes the same thing.

File: src/db/transaction.rs [hunk 9]
```diff
@@ -77,6 +77,10 @@ impl DbTransactionPool {
         }
     }
 
+    pub async fn get_pool(&self) -> Result<Box<dyn DbPool>, Error> {
+        Ok(self.pool.clone())
+    }
```
Comment (Mcat12) [inline src/db/transaction.rs:82]: This doesn't need to be an async function.

File: src/db/mod.rs [hunk 10]
```diff
@@ -186,8 +186,8 @@ pub trait Db<'a>: Debug + 'a {
             None => {
                 return Box::pin(
                     self.get_collection_timestamp(params::GetCollectionTimestamp {
-                        user_id,
-                        collection,
+                        user_id: user_id,
+                        collection: collection,
```
Comment (Mcat12) [inline src/db/mod.rs]: This should trigger a lint?

File: src/db/transaction.rs [hunk 11]
```diff
@@ -122,7 +131,36 @@ impl DbTransactionPool {
...
-        let (resp, db) = self.transaction_internal(action).await?;
+        // Get connection from pool
+        let db = self.pool.get().await?;
+        let resource_ts = db
+            .extract_resource(
+                self.user_id.clone(),
+                self.collection.clone(),
+                self.bso_opt.clone(),
+            )
+            .await?;
```
Comment (Mcat12) [inline src/db/transaction.rs:142]: This should probably happen inside of the transaction.

Comment (fzzzy) [reply src/db/transaction.rs:142]: Ah yes. Thank you!

File: src/db/spanner/manager.rs [hunk 12]
```diff
@@ -70,24 +88,17 @@ impl ManageConnection for SpannerConnectionManager {
...
-            match e {
-                grpcio::Error::RpcFailure(ref status)
-                    if status.status == grpcio::RpcStatusCode::NOT_FOUND =>
-                {
-                    conn.session = create_session(&conn.client, &self.database_name)?;
-                }
-                _ => return Err(e),
-            }
+        if let Err(e) = conn.client.get_session_async(&req)?.await {
+            return Err(e);
```
[the remark this answers is not in the document]
Comment (fzzzy) [reply src/db/spanner/manager.rs]: Thanks, I re-added it.

File: src/db/transaction.rs [hunk 13]
```diff
@@ -77,12 +81,17 @@ impl DbTransactionPool {
...
+    }
+
     /// Perform an action inside of a DB transaction.
-    pub async fn transaction<'a, A, R, F>(&'a self, action: A) -> Result<R, Error>
+    pub async fn transaction<'a, A: 'a, R, F>(&'a self, action: A) -> Result<R, Error>
     where
         A: FnOnce(Box<dyn Db<'a>>) -> F,
         F: Future<Output = Result<R, Error>> + 'a,
     {
+        // Get connection from pool
```
Comment (Mcat12) [inline src/db/transaction.rs:94]: nit: This comment is now out of place (and below, in `transaction_http`).

File: src/db/transaction.rs [hunk 14]
```diff
@@ -131,7 +187,7 @@ impl DbTransactionPool {
             self.report_error(&e);
             return Err(e.into());
         }
-
+
```
Comment (Mcat12) [inline src/db/transaction.rs:190]: nit: rustfmt might not like this.

File: src/db/spanner/manager.rs (hunk 2 above)
Comment (Mcat12) [reply src/db/spanner/manager.rs:98]: This got reverted back to `get_session`, needs a re-revert or something back to `get_session_async`.

File: src/db/spanner/manager.rs [hunk 15]
```diff
@@ -70,14 +88,14 @@ impl ManageConnection for SpannerConnectionManager {
...
+            use_test_transactions: self.test_transactions,
         })
     }
 
-    fn is_valid(&self, conn: &mut Self::Connection) -> Result<(), Self::Error> {
+    async fn is_valid(&self, mut conn: Self::Connection) -> Result<Self::Connection, Self::Error> {
         let mut req = GetSessionRequest::new();
         req.set_name(conn.session.get_name().to_owned());
-        if let Err(e) = conn.client.get_session(&req) {
+        if let Err(e) = conn.client.get_session_async(&req) {
```
Comment (Mcat12) [inline src/db/spanner/manager.rs:98]: Probably needs `.await`?

Comment (Mcat12) [reply src/db/spanner/manager.rs:98]:
```suggestion
        if let Err(e) = conn.client.get_session_async(&req)?.await {
```

File: src/db/transaction.rs [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+        if let Err(e) = result {
+            self.report_error(&e);
+            return Err(e.into());
+        }
+
+        Ok(resp)
+    }
+
+    /// Report an error to sentry, if possible
+    fn report_error(&self, error: &ApiError) {
```
[the remark this answers is not in the document]
Comment (fzzzy) [reply src/db/transaction.rs:110]: https://github.com/mozilla-services/syncstorage-rs/issues/726

File: src/db/transaction.rs (hunk 13 above)
Comment (fzzzy) [reply src/db/transaction.rs:94]: https://github.com/mozilla-services/syncstorage-rs/issues/727

File: src/db/transaction.rs [hunk 17]
```diff
@@ -0,0 +1,285 @@
...
+            }
+        }
+    }
+
+    pub fn get_pool(&self) -> Result<Box<dyn DbPool>, Error> {
+        Ok(self.pool.clone())
+    }
+
+    /// Perform an action inside of a DB transaction.
+    pub async fn transaction<'a, A: 'a, R, F>(&'a self, action: A) -> Result<R, Error>
```
[the remark this answers is not in the document]
Comment (fzzzy) [reply src/db/transaction.rs:89]: https://github.com/mozilla-services/syncstorage-rs/issues/728

File: src/web/handlers.rs [hunk 18]
```diff
@@ -16,389 +17,438 @@ use crate::web::{X_LAST_MODIFIED, X_WEAVE_NEXT_OFFSET, X_WEAVE_RECORDS};
...
-            let mut resp = json!({
-                "success": success,
-                "failed": failed,
-            });
-
-            if !breq.commit {
-                resp["batch"] = json!(&id);
-                return Either::Left(future::ok(HttpResponse::Accepted().json(resp)));
-            }
+        .and_then(|_| future::ok(()))
```
Comment (pjenvey) [inline src/web/handlers.rs:325]: nit: this should be killed

**Comment by Mcat12** [comment]
I labeled all of the follow-up issues with [`bb8-followup`](https://github.com/mozilla-services/syncstorage-rs/issues?q=is%3Aissue+is%3Aopen+label%3Abb8-followup)

**Pull request merged**